### ISSUE TITLE
ExchangeServer.GetDelegates fails on some O365 mailboxes with the exception "The XML document ended unexpectedly."

### ISSUE THREAD

**Issue by BillArmstrong** [OfficeDev/ews-managed-api#121]
ExchangeServer.GetDelegates has started failing on some O365 mailboxes with the exception "The XML document ended unexpectedly."

Upon examining the responses received from O365 I've determined that the call either succeeds or fails depending on the version of the server that is responding to the call. If the server version is 15.1.1199.19, the call succeeds. If the server version is 15.1.1220.14, the call fails.

Here is a sample of a successful response.

```
<?xml version="1.0" encoding="utf-8"?>
<s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelope/">
  <s:Header>
    <h:ServerVersionInfo MajorVersion="15" MinorVersion="1" MajorBuildNumber="1199" MinorBuildNumber="19" Version="V2017_04_14" xmlns:h="http://schemas.microsoft.com/exchange/services/2006/types" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"/>
  </s:Header>
  <s:Body>
    <m:GetDelegateResponse ResponseClass="Success" xmlns:m="http://schemas.microsoft.com/exchange/services/2006/messages" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:t="http://schemas.microsoft.com/exchange/services/2006/types">
      <m:ResponseCode>NoError</m:ResponseCode>
      <m:DeliverMeetingRequests>DelegatesOnly</m:DeliverMeetingRequests>
    </m:GetDelegateResponse>
  </s:Body>
</s:Envelope>

```
Here is a sample of an unsuccessful response.

```
<?xml version="1.0" encoding="utf-8"?>
<s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelope/">
  <s:Header>
    <h:ServerVersionInfo MajorVersion="15" MinorVersion="1" MajorBuildNumber="1220" MinorBuildNumber="14" Version="V2017_04_14" xmlns:h="http://schemas.microsoft.com/exchange/services/2006/types" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"/>
  </s:Header>
  <s:Body>
    <GetDelegateResponse ResponseClass="Success" xmlns="http://schemas.microsoft.com/exchange/services/2006/messages" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
      <ResponseCode>NoError</ResponseCode>
      <ResponseMessages/>
      <DeliverMeetingRequests>DelegatesOnly</DeliverMeetingRequests>
    </GetDelegateResponse>
  </s:Body>
</s:Envelope>

```
Note the addition of the ResponseMessages element. I don't know if this is the cause of the exception, but it is the most notable difference between the two.

**Comment by davster** [comment]
Thanks for reporting this.  This does appear to be a regression on our side.  Based on the schema definition, if there are no delegates to return, the ResponseMessages collection should be omitted since the definition of ArrayOfDelegateUserResponseMessageType indicates that the minOccurs of each response messages is 1.  Fixing now.

Of interest, if you do have a delegate assigned, does the call succeed?

**Comment by gscales** [comment]
If you have Delegate it also now is returning differently eg

  <?xml version="1.0" encoding="utf-8"?>
  <s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelope/">
    <s:Header>
      <h:ServerVersionInfo MajorVersion="15" MinorVersion="1" MajorBuildNumber="1199" MinorBuildNumber="21" Version="V2017_04_14" xmlns:h="http://schemas.microsoft.com/exchange/services/2006/types" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" />
    </s:Header>
    <s:Body>
      <m:GetDelegateResponse ResponseClass="Success" xmlns:m="http://schemas.microsoft.com/exchange/services/2006/messages" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:t="http://schemas.microsoft.com/exchange/services/2006/types">
        <m:ResponseCode>NoError</m:ResponseCode>
        <m:ResponseMessages>
          <m:DelegateUserResponseMessageType ResponseClass="Success">
            <m:ResponseCode>NoError</m:ResponseCode>
            <m:DelegateUser>
              <t:UserId>
                <t:SID>S-1-5-21-653873357-1868611527-989924688-6770850</t:SID>
                <t:PrimarySmtpAddress>timp19@datarumble.com</t:PrimarySmtpAddress>
            </m:DelegateUser>
          </m:DelegateUserResponseMessageType>
        </m:ResponseMessages>
        <m:DeliverMeetingRequests>NoForward</m:DeliverMeetingRequests>
      </m:GetDelegateResponse>
    </s:Body>
  </s:Envelope>
 
in 1220 you get
 
  <?xml version="1.0" encoding="utf-8"?>
  <s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelope/">
    <s:Header>
      <h:ServerVersionInfo MajorVersion="15" MinorVersion="1" MajorBuildNumber="1220" MinorBuildNumber="13" Version="V2017_04_14" xmlns:h="http://schemas.microsoft.com/exchange/services/2006/types" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" />
    </s:Header>
    <s:Body>
      <GetDelegateResponse ResponseClass="Success" xmlns="http://schemas.microsoft.com/exchange/services/2006/messages" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
        <ResponseCode>NoError</ResponseCode>
        <ResponseMessages>
          <GetDelegateResponseMessage ResponseClass="Success">
            <ResponseCode>NoError</ResponseCode>
            <DelegateUser>
              <UserId xmlns="http://schemas.microsoft.com/exchange/services/2006/types">
                <SID>S-1-5-21-4022512332-1284272732-3800139143-2063355</SID>
.....
            </DelegateUser>
          </GetDelegateResponseMessage>
        </ResponseMessages>
        <DeliverMeetingRequests>NoForward</DeliverMeetingRequests>
      </GetDelegateResponse>
    </s:Body>
  </s:Envelope>

So 1220 its returning GetDelegateResponseMessage instead of the normal DelegateUserResponseMessageType

**Comment by davster** [comment]
Yep.  I fixed Add, Get, Update and RemoveDelegates.  I verified both the no result and the result case and the element naming, etc... I will start a backport to some of the older branches to ensure that we get this fix out WW asap.  Sorry for the inconvenience.

**Comment by keshy** [comment]
I am seeing this too. This is breaking our prod. Thanks @davster - can you provide timelines for this fix on your side. 

Thanks

**Comment by davster** [comment]
Fixed in build 15.01.1265.000.  Getting approval for backport.  With a backport, we are talking a few days (say Monday-ish).  Without a backport, probably 2 weeks out.

**Comment by keshy** [comment]
Thanks @davster that helps to evaluate our options.

**Comment by itchy92** [comment]
Hi @davster, any update on the rollout of this fix? We modified our code to workaround this behavior, and wondering if we can revert back.

Thanks!
